src/System.Text.JsonLab/System/Text/Json/Utf8JsonReaderInstrumented_MultiSegment.cs

```diff
@@ -0,0 +1,460 @@
+                {
+                    if (first != JsonConstants.Quote)
+                        JsonThrowHelper.ThrowJsonReaderException(ref this);
+
+                    reader.Advance(1);
+                    CurrentIndex++;
+                    TokenStartIndex++;
+                    ConsumePropertyName(ref reader);
+                }
+                else
+                {
+                    ConsumeValue(ref reader, first);
+                }
+            }
+            else if (marker == JsonConstants.CloseBrace)
+            {
+                EndObject();
+            }
+            else if (marker == JsonConstants.CloseBracket)
+            {
+                EndArray();
+            }
+            else
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+        }
+
+        private void ConsumeValue(ref BufferReader<byte> reader, byte marker)
+        {
+            TokenType = JsonTokenType.Value;
+
+            if (marker == JsonConstants.Quote)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                TokenStartIndex++;
+                ConsumeString(ref reader);
+            }
+            else if (marker == JsonConstants.OpenBrace)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                StartObject();
+                ValueType = JsonValueType.Object;
+            }
+            else if (marker == JsonConstants.OpenBracket)
+            {
+                reader.Advance(1);
+                CurrentIndex++;
+                StartArray();
+                ValueType = JsonValueType.Array;
+            }
+            else if ((uint)(marker - '0') <= '9' - '0')
+            {
+                ConsumeNumber(ref reader);
+            }
+            else if (marker == '-')
+            {
+                //TODO: Is this a valid check or do we need to do an Advance to be sure?
+                if (reader.End) JsonThrowHelper.ThrowJsonReaderException(ref this);
+                ConsumeNumber(ref reader);
+            }
+            else if (marker == 'f')
+            {
+                ConsumeFalse(ref reader);
+            }
+            else if (marker == 't')
+            {
+                ConsumeTrue(ref reader);
+            }
+            else if (marker == 'n')
+            {
+                ConsumeNull(ref reader);
+            }
+            else if (marker == '/')
+            {
+                // TODO: Comments?
+                JsonThrowHelper.ThrowNotImplementedException();
+            }
+            else
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+        }
+
+        private void ConsumeNumber(ref BufferReader<byte> reader)
+        {
+            //TODO: Increment Index by number of bytes advanced
+            if (!reader.TryReadToAny(out ReadOnlySpan<byte> span, JsonConstants.Delimiters, advancePastDelimiter: false))
```

Comment:
@JeremyKuhne, would it be possible for TryReadToAny to return the entire data as an output span if none of the delimiters are found? The fact that this returns an empty span when no delimiters are found that results in me having to special case the following:
https://github.com/dotnet/corefxlab/pull/2493/files#diff-15d23dd0f42f71cf1a5e5426e7a2c609R66

Reply:
It could, but I think that would potentially be confusing and, more importantly, allocate unnecessarily. Why doesn't ConsumeNumber contain this logic?

Reply:
Why would it allocate?

Reply:
> Why would it allocate?

Because he wants the rest of the ROS as a span- at least that is what the code does here.

Reply:
Ah, I just noticed that the span is an out parameter. We should talk how we want to read/parse variable length values. Such out parameter will allocate in a lot of cases and I think we need a way to do it in a non-allocating way. When I initially prototypes BufferReader, I think I would return Position and/or int from methods that read variable size values, e.g. strings.